runtime/near-vm-runner/RUNTIMES.md

```diff
@@ -0,0 +1,214 @@
+target is a hard requirement.
+
+However, the contract developers may not necessarily have a `x86_64` machine to develop with, or
+they might be using a device running a different operating system. With the recent surge in
+popularity of devices based on the Arm CPU architecture chips and generally high popularity of the
+macOS and Windows operating systems it would be a huge boon to their development experience if the
+runtime used also supported these other targets.
+
+## Performance
+
+There are a couple different ways we can evaluate the performance of the runtime. How quickly can
+the runtime execute the instructions (throughput), execute the first instruction (latency), prepare
+the initial code for execution (compilation) and how efficient are miscellaneous operations such as
+serialization and deserialization of the module.
+
+In order to realistically evaluate the performance of a runtime it is important to understand how
+the NEAR validator utilizes the runtime to implement its functionality. Somewhat simplified view
+is:
+
+1. A module containing Wasm code in binary format is parsed and instrumented;
+2. Wasm code produced after instrumentation is supplied to the runtime for preparation
+   (compilation);
+3. The prepared module is serialized and written to a long term storage (database);
+4. When the contract function call occurs, the serialized module is deserialized, loaded into
+   memory and the execution of the desired function begins (latency);
+5. The contract code is executed (throughput) and the results are returned to the validator.
+
+It is important to remember that the NEAR protocol charges fees for an operation before the
+operation is executed. For that reason, predictability of worst case execution time often matters
+more than the execution time in the typical case. It is important that we are able to deduce ahead
+of time what cost to assign to a given operation. Inability to do so can lead to significant
+undercharging and break the properties underlying the protocol.
+
+On a scale of trade-offs, performance is probably one of the less important metrics. Slower
+execution of a contract code doesn't make NEAR protocol unsound as an idea and only serves to
+impose stricter limits on what can be achieved on the network.
+
+### Compilation
+
+The NEAR protocol charges gas fees to deploy a contract. One part of a deployment involves
+preparing the contract Wasm code for execution. This may involve instrumentation, compilation,
+serialization of the machine code and other similar operations.
+
+An estimate of the deployment cost will typically involve a function which uses the size of the
+input Wasm code as its primary input. Such a function can only exist if we have a good knowledge of
+the runtime’s time complexity properties are known. For our purposes a linear or `O(n log n)`
+relationship between the input size and execution time is the highest we can accept.
```

Comment:
Actually, what is the reason why a quadratic algorithm is absolutely unacceptable? @matklad

Reply:
The max size of the contract is 4mb, so we can have up to roughly `10^6` "things" (functions, instruction in a single function, totall call depth, etc). `10^6 ^ 2` would be prohibitively expensive to compute. 

That being said, we *could* accept wore time complexities if we reflect them in the cost model. For example, if we now that our register allocation is, eg, quadratic in the number of local variables, we can charge quadratic cost for that. Non-linear cost models feel like the can of worms though.

Reply:
I think it could be feasible to utilize quadratic algorithms for select stages of our runtime if we did meter and charge for those stages separately from everything else. That way the linear operations with larger constant factors would be isolated from the quadratic stuff with potentially small factors and either approach would be reasonably feasible.

That said, since we're trying to estimate fees from the input wasm code, establishing the relationship between input wasm and the work an algorithm such as regalloc would do would probably involve a fair amount of tessomancy.